core/src/main/scala/scalaz/EitherT.scala

```diff
@@ -287,6 +287,39 @@ object EitherT extends EitherTInstances {
 
 }
 
+object EitherTNondeterminism {
+  implicit def eitherTNondeterminism[F[_], L](implicit F0: Nondeterminism[F]): Nondeterminism[EitherT[F, L, ?]] =
+    new Nondeterminism[EitherT[F, L, ?]] {
+      val F = Nondeterminism[F]
+
+      override def chooseAny[A](
+                                 head: DisjunctionT[F, L, A],
+                                 tail: Seq[DisjunctionT[F, L, A]]): DisjunctionT[F, L, (A, Seq[DisjunctionT[F, L, A]])] =
+        new EitherT(F.map(F.chooseAny(head.run, tail map (_.run))) {
+          case (a, residuals) =>
+            a.map((_, residuals.map(new EitherT(_))))
+        })
+
+      override def bind[A, B](fa: DisjunctionT[F, L, A])(f: (A) => DisjunctionT[F, L, B]): DisjunctionT[F, L, B] =
+        fa flatMap f
+
+      override def point[A](a: => A): DisjunctionT[F, L, A] = EitherT.eitherTMonad[F, L].point(a)
+
+      override def reduceUnordered[A, M](fs: Seq[DisjunctionT[F, L, A]])(
+        implicit R: Reducer[A, M]): DisjunctionT[F, L, M] = {
+        import R.monoid
+        val AE = Apply[L \/ ?]
+        val RR: Reducer[L \/ A, L \/ M] =
+          Reducer[L \/ A, L \/ M](
+            _.map(R.unit),
+            c => AE.apply2(c, _)(R.cons(_, _)),
+            m => AE.apply2(m, _)(R.snoc(_, _))
+          )(Monoid.liftMonoid[L \/ ?, M])
+        new EitherT(F.reduceUnordered(fs.map(_.run))(RR))
+      }
```

Comment:
Unfortunately, `Task` can get away with it, since it knows the details of the underlying functor; `EitherT` does not, so cannot apply the exact same technique.

The default implementation is a principled, if not optimally performant, way to get the desired behavior. For `EitherT`'s instance to do the same, @izhangzhihao will need to come up with a principled way to expose the needed information in `Nondeterminism[F]` so the `EitherT` instance can universally preserve the behavior. I don't imagine this will look like "expose a convenient isomorphism to Either", because that way surely lies uninstantiability.

Sorry to be so vague. Solving this properly will be tricky, and I can't exactly explain what needs to be done.